# Linear Regression Practice

##### For this exercise, you've been provided with a csv file, gdp_le.csv, which contains the gdp per capita and life expectancy values that you were working with on the UN data exercise.


In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import statsmodels.formula.api as smf
import statsmodels.stats.api as sms
import scipy as sp

### 1. Start by fitting a linear regression model with target being life expectancy and predictor variable year.  

To specify our model, we need to use a [patsy formula](https://patsy.readthedocs.io/en/latest/formulas.html). The recipe is `target variable ~ predictor variables`.

We'll use the `ols` function and go ahead and fit the model. (ols = ordinary least squares)

In [2]:
gdp_le = pd.DataFrame(pd.read_csv('../data/gdp_le.csv'))
year_vs_le = smf.ols("Life_Expectancy ~ Year", data=gdp_le).fit()

####    a. What coefficients do you get? Interpret the meaning of these coefficents.  

In [3]:
year_vs_le.params

Intercept   -468.138140
Year           0.267722
dtype: float64

- *the formula would fill as: $$Mean(Life\_Expectancy) = -468.138140 + 0.267722(Year)$$ meaning each increase in year on average the Life Expectancy increases by ~0.267722 years*

####    b. Using just the year makes the intercept term difficult to interpret. Create a new model, but this time use years since 1990 as your predictor variable. Hint: You can use the [patsy identity function](https://patsy.readthedocs.io/en/latest/builtins-reference.html#patsy.builtins.I) to modify your predictors in your model formula. Inspect the new coefficients and interpret the meaning of them. Are they statistically significant?  

In [11]:
gdp_le = gdp_le.sort_values('Year').set_index('Year')
year1990_vs_le = smf.ols('Life_Expectancy ~ gdp_le.index', data=gdp_le).fit()
year1990_vs_le.params

Intercept      -468.138140
gdp_le.index      0.267722
dtype: float64

####    c. Compare the actual mean life expectancy per year to the model's estimate. How well does it do?  

####    d. Plot the actual values against your model's estimates for the mean life expectancy.  

####    e. Inspect the R-squared value for the model. does it make sense, given the plot?  

### 2. Filter the full dataset down to just the year 2021. Fit a linear regression model with target being life expectancy and predictor variable gdp per capita.  

    a. What coefficients do you get? Interpret the meaning of those coefficients.  

    b. Refit your model, but this time use thousands of dollars of gdp per capita as your predictor. How does this change your coefficients?  

    c. Are the coefficients statistically significant?  

    d. What does your model estimate for the mean life expectancy for a country whose gdp per capita is $50,000? What about one whose gdp per capita is $100,000?

    e. Plot the actual values compared to your model's estimates for mean life expectancy. How would you assess the model's fit

### 3. Now, fit a model for life expectancy based on the log of gdp per capita.  

    a. Inspect the coefficients for this model. Are they statistically significant?

    b. Interpret these coefficients. What does the model estimate for the average life expectancy for countries with a gdp per capita of $50,000? What about for those with a gdp per capita of $100,000?

    c. Plot the actual values compared to your models' estimates for the mean life expectancy. How does this compare to the non-logged model?

### 4. Finally, return to the full dataset. 

    a. First, fit a linear regression model for life expectancy based on the log of gdp per capita.

    b. Then, add the year variable to your model. How can you interpret the coefficient associated with year? What limitations or weaknesses might this model have?